### ISSUE TITLE
Bugs in equation evaluation, dataset generation, and skeleton optimization

### ISSUE THREAD

**Issue by t-taniai** [mojivalipour/symbolicgpt#1]
Thank you very much for sharing the code.

I'm carefully studying the code and found a couple of problems that the authors may have not intended to happen.

1. Issue in evaluating an equation using `eval`
Assume an equation skeleton is "C*x1**2" and C=1 and x1=-2.0. Its correct value is '4'. In the code, it is evaluated by `eval("1.0*-2.0**2")`, which however is interpreted as `1.0*-(2.0**2)` and results in `-4`. This bug affects the dataset generation (as values of Y) as well as the test score evaluation where `eval` is also used (in computations of `lossFunc` and Ys).

2. Issue in optimizing skeleton coefficients
The second problem is in the optimization of C's in an estimated skeleton sing `scipy.optimize.minimize`. In its loss function defined as `utils.lossFunc`,  `relativeErr` is used to evaluate a loss between Y and Yhat. However, relativeErr is called for each pair of elements of Y and Yhat. So, instead of optimizing `mean( (Y-Yhat)**2/norm(Y) )`, it actually optimizes `mean( (Y-Yhat)**2/abs(Y) )`. Since the final score is evaluated based on `(Y-Yhat)**2/norm(Y)`, I think it is more natural to optimize the equations based on the same metric.

3. Datasets contain invalid equations
The datasets contain invalid equations such as `sqrt(-1.0*abs(x1))` and `log(-1.0*abs(x1))`. There are lines in the code to reject those invalid equations during generation. But it is not working because `eval` function in `generateDataStrEq`  refers to safe wrappers of sqrt and log rather than numpy's functions themselves. So, even if an equation is invalid, it produces no NaN nor Inf and is hence not rejected.

Probably, the first and third issues can be solved by using a dict augment of eval.
```
ref_dict = {
    'x1': -1,
    'sqrt': numpy.sqrt
}
y = eval('sqrt(x1**2)', ref_dict)
```
This way, for the first issue, `sqrt(x1**2)` is properly evaluated as sqrt((-1)**2). You can also 'eval' a numpy.ndarray as `x1` directly without looping a list of floats. (Btw, I did this and found that ndarray needs to be numpy.float64 to get enough precision to run `scipy.optimize.minimize`.)

As for the thrid issue, you can use this technique to call numpy funcs in eval in generateDataStrEq. Then, invalid equations will produce nan and inf to be properly rejected by your validity check code.

Thanks,
Tatsunori

**Comment by t-taniai** [comment]
Hi, I am preparing bug fixes for these issues and a PR.
Currently, I'm fixing the dataset generation code. To be complete, I want to get `config.txt` for the dataset named `1-9Var_RandSupport_FixedLength_-3to3_-5.0to-3.0-3.0to5.0_20-250`, for which the data generation procedure is missing. 
What are the parameters for this dataset? And what are `Test_Benchmarks` and `Test_General` subsets and how to produce them?
Thanks!

**Comment by t-taniai** [comment]
Btw, in some datasets (1Var, 2Var, and 3Var), `sqrt` is missing in `op_list` parameter. Is this intended?

**Comment by mojivalipour** [comment]
@t-taniai Basically, we wanted to complicate things for the general model.

> Btw, in some datasets (1Var, 2Var, and 3Var), `sqrt` is missing in `op_list` parameter. Is this intended?

**Comment by mojivalipour** [comment]
The 1-9Var_RandSupport_FixedLength_-3to3_-5.0to-3.0-3.0to5.0_20-250 config.txt has been added. 

> Hi, I am preparing bug fixes for these issues and a PR. Currently, I'm fixing the dataset generation code. To be complete, I want to get `config.txt` for the dataset named `1-9Var_RandSupport_FixedLength_-3to3_-5.0to-3.0-3.0to5.0_20-250`, for which the data generation procedure is missing. What are the parameters for this dataset?

**Comment by mojivalipour** [comment]
The Test_General is the same as the other experiments. This set has the same distribution as the training set. In Test_Benchmark, however, there are only benchmark equations. Our paper will be updated shortly with additional details. 

> And what are `Test_Benchmarks` and `Test_General` subsets and how to produce them? Thanks!
